### PR TITLE
#223 dynamic wiremock ports

### LINKED ISSUE #193: Change WireMock port
We are currently using the `8080` port for our `WireMock` tests.
This port seems to be conflicting with Tomcat, Quarkus and Spring-Boot default ports.
We need to find an alternative port.

### LINKED ISSUE #223: Alter WireMock tests to use dynamic ports.
Currently, the ports are hardcoded. Use dynamic ports. For this, the url files in _ide/urls/... must be created during the test runtime using the assigned dynamic port.

Tests which currently use wiremock:

**cli/src/test/java/com/devonfw/tools/ide/tool/jmc/JmcTest.java

cli/src/test/java/com/devonfw/tools/ide/commandlet/InstallCommandletTest.java**

### PR DESCRIPTION
**Pull request by slskiba**
Closes #223 and closes #193

### CODE REVIEW

**Comment by slskiba** [comment]
Also closes #193

File: cli/src/test/java/com/devonfw/tools/ide/tool/UrlUpdaterMock.java
Comment (jan-vcapgemini) [inline cli/src/test/java/com/devonfw/tools/ide/tool/UrlUpdaterMock.java:22]: Please add some Javadoc to please my IDE ;)

File: cli/src/test/java/com/devonfw/tools/ide/tool/UrlUpdaterMockSingle.java
Comment (jan-vcapgemini) [inline cli/src/test/java/com/devonfw/tools/ide/tool/UrlUpdaterMockSingle.java:18]: Please add some Javadoc to please my IDE ;)

File: cli/src/test/java/com/devonfw/tools/ide/tool/androidstudio/AndroidStudioUrlUpdaterMock.java
Comment (jan-vcapgemini) [inline cli/src/test/java/com/devonfw/tools/ide/tool/androidstudio/AndroidStudioUrlUpdaterMock.java:13]: Please add some Javadoc to please my IDE ;)

File: cli/src/test/java/com/devonfw/tools/ide/tool/intellij/IntellijUrlUpdaterMock.java
Comment (jan-vcapgemini) [inline cli/src/test/java/com/devonfw/tools/ide/tool/intellij/IntellijUrlUpdaterMock.java:12]: Please add some Javadoc to please my IDE ;)

File: cli/src/test/java/com/devonfw/tools/ide/tool/pip/PipUrlUpdaterMock.java
Comment (jan-vcapgemini) [inline cli/src/test/java/com/devonfw/tools/ide/tool/pip/PipUrlUpdaterMock.java:19]: Please add some Javadoc to please my IDE ;)

File: cli/src/test/java/com/devonfw/tools/ide/tool/python/PythonUrlUpdaterMock.java
Comment (jan-vcapgemini) [inline cli/src/test/java/com/devonfw/tools/ide/tool/python/PythonUrlUpdaterMock.java:12]: Please add some Javadoc to please my IDE ;)

**Review by jan-vcapgemini — APPROVED** [review]
Nice fix. Finally we can use dynamic ports for WireMock.
I've added some very small CRs (just to please my IDE).

**Review by jan-vcapgemini — APPROVED** [review]
Thanks, ready for review.

**Comment by slskiba** [comment]
PR #433 will bring new tests with fixed ports, that should also be adjusted as part of this PR.

**Review by hohwille — APPROVED** [review]
@slskiba thanks for your PR. This looks really good to me and is IMHO ready for merge. Nice job 👍 
Do you have some known issues I am missing or why is this PR still in draft mode? Build already succeeded so you did not break the wiremock tests and now they should also work if port `8080` is already used by another app what is often the case on my machine...

**Comment by slskiba** [comment]
@hohwille as stated above, this PR is waiting for the merge of #433 as it brings new tests with fixed ports, so that they can also be immediately adjusted to use dynamic ports before merging this.

**Comment by hohwille** [comment]
> @hohwille as stated above, this PR is waiting for the merge of #433 as it brings new tests with fixed ports, so that they can also be immediately adjusted to use dynamic ports before merging this.

Is merged meanwhile. I just updated this branch/PR so you should have the TODO if you pull and can resolve this last spot as well and then this should be ready.
Thanks for explanation and taking care of everything 👍

File: cli/src/test/java/com/devonfw/tools/ide/tool/intellij/IntellijTest.java
Comment (hohwille) [inline cli/src/test/java/com/devonfw/tools/ide/tool/intellij/IntellijTest.java:42]: a golden rule for JUnits is that they should never do mutations that would be tracked by git.
Wouldn't it be possible and even easier to simply write the file in target relative to `context` in `setupMockedPlugin()`?

File: cli/src/test/java/com/devonfw/tools/ide/tool/intellij/IntellijJsonUrlUpdaterTest.java
Comment (hohwille) [inline cli/src/test/java/com/devonfw/tools/ide/tool/intellij/IntellijJsonUrlUpdaterTest.java:57]: Here we also write dynamic content to files.
Great thing here is that you are using `@TempDir` so it is fine.
However, writing a file in `@BeforeAll` to disc just for the sake to later read it in the test from the disk again seems also kind of pointless.
Instead of using the `@TempDir` you could simply store the content as String in a variable and take it from there instead of using the disc for communication.
There are some other places with the same pattern.
Surely a nice-to-have improvement but would still make sense to me to improve it...

**Review by hohwille — COMMENTED** [review]
@slskiba thanks for completing. 👍 
I found a small problem and being pointed to that I also left a second comment with a similar pattern that is not wrong but should be improved.
Can you please again have a look? Thanks in advance!

**Review by hohwille — APPROVED** [review]
@slskiba fast and perfect addressing of my review comments. Great job 🚀

**Pull request merged**
